### PR TITLE
Move symex state levels to a new file

### PR DESCRIPTION
**Pull request by romainbrenguier**
No functional change. 
The goal is to reduce the complexity of goto_symex_state by extracting a module which takes care of the renaming levels. 

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [na] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f08782b93619bfc8fd0bfb1981fe645148e7b786).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91673552

File: src/goto-symex/renaming_level.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+#ifndef CPROVER_GOTO_SYMEX_RENAMING_LEVEL_H
+#define CPROVER_GOTO_SYMEX_RENAMING_LEVEL_H
+
+#include <map>
+#include <unordered_set>
+
+#include <util/irep.h>
+#include <util/ssa_expr.h>
+
+struct renaming_levelt
```
Comment (peterschrammel) [inline src/goto-symex/renaming_level.h]: Would be good to add also some documentation at this occasion.

**Review by peterschrammel — APPROVED** [review]

File: src/goto-symex/renaming_level.h (hunk 1 above)
Comment (romainbrenguier) [reply src/goto-symex/renaming_level.h]: I added some documentation now.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: be65d1599308ea7eb1f4aed3b986cbc1b63a9841).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91841291

File: src/goto-symex/renaming_level.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+{
+  virtual ~renaming_levelt() = default;
```
Comment (antlechner) [inline src/goto-symex/renaming_level.h]: :pick: It would have been helpful to have the moving part and making changes in separate commits.

File: src/goto-symex/renaming_level.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  void increase_counter(const irep_idt &identifier)
+  {
+    PRECONDITION(current_names.find(identifier) != current_names.end());
+    ++current_names[identifier].second;
+  }
+
+  void get_variables(std::unordered_set<ssa_exprt, irep_hash> &vars) const
+  {
+    for(const auto &pair : current_names)
```
Comment (antlechner) [inline src/goto-symex/renaming_level.h:50]: :pick: Can you think of a more meaningful name than `pair`?

File: src/goto-symex/renaming_level.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+// level 0 -- threads!
+// renaming built for one particular interleaving
+struct symex_level0t : public renaming_levelt
+{
+  void
+  operator()(ssa_exprt &ssa_expr, const namespacet &ns, unsigned thread_nr);
+
+  symex_level0t() = default;
+  ~symex_level0t() override = default;
```
Comment (antlechner) [inline src/goto-symex/renaming_level.h:64]: :question: Why is this line needed? :horse:

File: src/goto-symex/renaming_level.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      {
+        PRECONDITION(it->first == pair.first);
+        it->second = pair.second;
+        ++it;
+      }
+    }
+  }
+
+  symex_level1t() = default;
+  ~symex_level1t() override = default;
```
Comment (antlechner) [inline src/goto-symex/renaming_level.h:78]: :horse:

File: src/goto-symex/renaming_level.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+};
+
+// level 2 -- SSA
+struct symex_level2t : public renaming_levelt
+{
+  symex_level2t() = default;
+  ~symex_level2t() override = default;
```
Comment (antlechner) [inline src/goto-symex/renaming_level.h:87]: :horse:

File: src/goto-symex/renaming_level.h [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  operator()(ssa_exprt &ssa_expr, const namespacet &ns, unsigned thread_nr);
+
+  symex_level0t() = default;
+  ~symex_level0t() override = default;
+};
+
+// level 1 -- function frames
+// this is to preserve locality in case of recursion
+
+struct symex_level1t : public renaming_levelt
```
Comment (antlechner) [inline src/goto-symex/renaming_level.h]: :question: Why are these renamed to `symex_level1t`, but the parent is not renamed to `symex_renaming_levelt`?

File: src/goto-instrument/accelerate/scratch_program.cpp [hunk 8]
```diff
@@ -35,7 +35,7 @@ bool scratch_programt::check_sat(bool do_slice)
   output(ns, "scratch", std::cout);
 #endif
 
-  goto_symex_statet::propagationt::valuest constants;
```
Comment (antlechner) [inline src/goto-instrument/accelerate/scratch_program.cpp:38]: :question: This doesn't actually seem to be used anywhere, can it be removed?

File: src/goto-symex/goto_symex_state.cpp [hunk 9]
```diff
@@ -298,15 +295,21 @@ void goto_symex_statet::rename(
...
     {
-      set_ssa_indices(ssa, ns, level);
+      set_l1_indices(ssa, ns);
       rename(expr.type(), ssa.get_identifier(), ns, level);
       ssa.update_type();
     }
     else if(level==L2)
     {
-      set_ssa_indices(ssa, ns, L1);
+      set_l1_indices(ssa, ns);
```
Comment (antlechner) [inline src/goto-symex/goto_symex_state.cpp:312]: :bulb: "If L2 then use the L1 version" is quite confusing. I see that it was the same before your changes, but if you know the reason, could you explain it here?

File: src/goto-symex/renaming_level.cpp [hunk 10]
```diff
@@ -48,7 +48,7 @@ void symex_level1t::operator()(ssa_exprt &ssa_expr)
 
   const irep_idt l0_name = ssa_expr.get_l1_object_identifier();
 
-  current_namest::const_iterator it = current_names.find(l0_name);
+  auto it = current_names.find(l0_name);
```
Comment (antlechner) [inline src/goto-symex/renaming_level.cpp]: :question: Shouldn't these be `const`?

File: src/goto-symex/renaming_level.h [hunk 11]
```diff
@@ -18,34 +18,41 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
...
   {
     for(const auto &pair : current_names)
       vars.insert(pair.second.first);
   }
 };
 
-// level 0 -- threads!
-// renaming built for one particular interleaving
+/// Functor to set the level 0 renaming of SSA expressions.
+/// Level 0 corresponds to threads.
```
Comment (antlechner) [inline src/goto-symex/renaming_level.h:56]: :+1:

File: src/goto-symex/renaming_level.h [hunk 12]
```diff
@@ -18,34 +18,41 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 #include <util/irep.h>
 #include <util/ssa_expr.h>
 
+/// Wrapper for a \c current_names map, which maps each identifier to an SSA
```
Comment (antlechner) [inline src/goto-symex/renaming_level.h:21]: :question: What do we use this map for? Could you explain a bit more about the context of this?

**Review by antlechner — COMMENTED** [review]
I'm not familiar with this part of the code but it's all refactoring and the changes look good to me! Just a few comments and questions before I can approve.

File: src/goto-symex/renaming_level.h (hunk 2 above)
Comment (romainbrenguier) [reply src/goto-symex/renaming_level.h]: Yes sorry about that I think I had a commit before the moving that I removed by mistake before rebasing

File: src/goto-symex/renaming_level.h (hunk 4 above)
Comment (romainbrenguier) [reply src/goto-symex/renaming_level.h:64]: There may subtleties I don't fully understand there, as to why there are needed. The constructor and destructor were defined before so I left the declaration, but then clion recommends to use `= default` instead of `{}` because it is more explicit (for the compiler in particular). Tagging @LAJW in case he can bring more insight here.

File: src/goto-instrument/accelerate/scratch_program.cpp (hunk 8 above)
Comment (romainbrenguier) [reply src/goto-instrument/accelerate/scratch_program.cpp:38]: :+1: good catch

File: src/goto-symex/goto_symex_state.cpp (hunk 9 above)
Comment (romainbrenguier) [reply src/goto-symex/goto_symex_state.cpp:312]: I agree it's a bit confusing and I don't know exactly how it works, but I think the idea is that doing the renaming for one level you should also take care of the lower levels (for level1 we also do level0 and for level2 we do level0 and level1)

**Comment by romainbrenguier** [comment]
@antlechner ready for rereview

File: src/goto-symex/renaming_level.h (hunk 4 above)
Comment (antlechner) [reply src/goto-symex/renaming_level.h:64]: It just looks a bit strange that we're overriding a method to do exactly what its parent method does, but if it was the same before then any decision to change it can be made in a future PR.

**Review by antlechner — APPROVED** [review]
LGTM

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b265f7f206aecdfae482238348fb1d15832366c7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92252168

File: src/goto-symex/goto_symex_state.cpp (hunk 9 above)
Comment (tautschnig) [reply src/goto-symex/goto_symex_state.cpp:312]: We do the L2 renaming below, but first need to make sure the expression has all L1 levels set correctly, in particular in the type.

**Review by tautschnig — APPROVED** [review]
I think this is good, and probably doesn't affect performance. Given SV-COMP deadlines, right now is either the perfect time to merge this or a very bad time.
